### PR TITLE
[SPARK-16803] [SQL] SaveAsTable does not work when source DataFrame is built on a Hive Table

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

In Spark 2.0, `SaveAsTable` does not work when source DataFrame is built on a Hive Table, but Spark 1.6 works.

**Spark 1.6**

``` Scala
scala> sql("create table sample.sample stored as SEQUENCEFILE as select 1 as key, 'abc' as value")
res2: org.apache.spark.sql.DataFrame = []

scala> val df = sql("select key, value as value from sample.sample")
df: org.apache.spark.sql.DataFrame = [key: int, value: string]

scala> df.write.mode("append").saveAsTable("sample.sample")

scala> sql("select * from sample.sample").show()
+---+-----+
|key|value|
+---+-----+
|  1|  abc|
|  1|  abc|
+---+-----+
```

**Spark 2.0**

``` Scala
scala> df.write.mode("append").saveAsTable("sample.sample")
org.apache.spark.sql.AnalysisException: Saving data in MetastoreRelation sample, sample
 is not supported.;
```

So far, we do not plan to support it in Spark 2.0. Spark 1.6 works because it internally uses `insertInto`. But, if we change it back it will break the semantic of `saveAsTable` (this method uses by-name resolution instead of using by-position resolution used by `insertInto`).

Instead, users should use `insertInto` API. This PR corrects the error messages. Users can understand how to bypass it before we support it in a separate PR.
### How was this patch tested?

Test cases are added

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @yhuai @rxin @liancheng @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala [hunk 1]
```diff
@@ -236,6 +236,11 @@ case class CreateDataSourceTableAsSelectCommand(
               existingSchema = Some(l.schema)
             case s: SimpleCatalogRelation if DDLUtils.isDatasourceTable(s.metadata) =>
               existingSchema = Some(DDLUtils.getSchemaFromTableProperties(s.metadata))
+            // When the source table is a Hive table, the `saveAsTable` API of DataFrameWriter
+            // does not work. Instead, use the `insertInto` API.
+            case o if o.getClass.getName == "org.apache.spark.sql.hive.MetastoreRelation" =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala]: it's really hacky to compare with the class name string...  What's the error message before?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala]: Agree! Do you have any idea how to improve it? 

```
org.apache.spark.sql.AnalysisException: Saving data in MetastoreRelation sample, sample
 is not supported.;
```

Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala]: `case o: MetastoreRelation` doesn't work?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala [hunk 2]
```diff
@@ -236,6 +236,11 @@ case class CreateDataSourceTableAsSelectCommand(
...
+              throw new AnalysisException("saveAsTable() for Hive tables is not supported yet. " +
```
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala]: Remove "yet" or point to a JIRA issue where it's going to be introduced.

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 3]
```diff
@@ -449,6 +449,22 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext with Be
...
 
+  test("saveAsTable(CTAS) when the source DataFrame is built on a SimpleCatalog table") {
+    import testImplicits._
+
+    val tableName = "tab1"
+    withTable(tableName) {
+      sql(s"CREATE TABLE $tableName(key int, value int) USING parquet")
+      (1 to 2).map { i => (i, i) }.toDF("key", "value").write.insertInto(tableName)
+      val df = sql(s"select key, value from $tableName")
+      df.write.mode("append").saveAsTable(tableName)
```
Comment (jaceklaskowski) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: What do you think about using `SaveMode.Append` instead?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 4]
```diff
@@ -1457,6 +1457,59 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
   }
 
+  test("saveAsTable(CTAS) when the source DataFrame is built on a Hive table") {
+    val tableName = "tab1"
+    withTable(tableName) {
+      sql(s"CREATE TABLE $tableName stored as SEQUENCEFILE as select 1 as key, 'abc' as value")
+      val df = sql(s"select key, value as value from $tableName")
+
+      val message = intercept[AnalysisException] {
+        df.write.mode("append").saveAsTable(tableName)
```
Comment (jaceklaskowski) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: and here too?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala]: `MetastoreRelation` is in another package `hive`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala]: ok, will remove it.

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: sure, we can do it.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala (hunk 4 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: ok

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 5]
```diff
@@ -1457,6 +1457,59 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
+        Row(1, 1) :: Row(2, 2) :: Row(3, 4) :: Row(3, 4) :: Nil
+      )
+    }
+  }
+
+  test("insertInto when the source DataFrame is built on a Hive table") {
+    val tableName = "tab1"
+    withTable(tableName) {
+      sql(s"CREATE TABLE $tableName stored as SEQUENCEFILE as select 1 as key, 'abc' as value")
+      val df = sql(s"select key, value as value from $tableName")
```
Comment (jaceklaskowski) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Is there a reason why you're using lowercase variant of the query in `sql` while `CREATE TABLE` above is uppercase? It seems inconsistent.

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Sure, we can make them consistent.

**Comment by gatorsmile** [comment]
@cloud-fan @yhuai  https://github.com/apache/spark/compare/master...gatorsmile:saveAsTableFix?expand=1

This is the prototype for supporting this missing function. We can fix it with minor changes. To avoid too many calls on `lookupRelation`, we might need to move the plan conversion to a separate batch. 

Let me know if we should do it in 2.0.1 or just issue an exception in 2.0.1. 

Thanks!

**Comment by cloud-fan** [comment]
now I think it's easy to catch `saveAsTable` for hive tables:
`case CreateTable(tableDesc, SaveMode.APPEND, Some(query)) if tableDesc.provider.get != "hive" =>`

We can add a new rule in hive module at issue meaningful error message.

**Comment by cloud-fan** [comment]
Hi @gatorsmile can you send out a PR first? We definitely need it for 2.1, and we can decide if we should backport it to 2.0 later.

**Comment by gatorsmile** [comment]
@cloud-fan Sure, will do it soon. Thanks!

**Comment by gatorsmile** [comment]
We can support it in another PR: https://github.com/apache/spark/pull/14612. This can be closed

**Pull request closed without merging**